node: refactor guardiand/node.go

tbjump opened the pull request:
Goals
* refactor the Guardian Node into its own package such that it can be tested through golang unit-tests.

Non-goals
* Any kind of refactoring beyond the stated goal is not in-scope. The interfaces of the various components could probably be further cleaned up and the handling of the configuration flags could probably be improved (e.g. with `viper`). 

High-level overview
* The `node` package is added with `type G`, where G stands for a Guardian Node. This allows creating multiple `G` in the unit tests. 
* Configuration of `G` is implemented through currying. This helps to separate the *specification* of the configuration from the *application* of the configuration. 
    * The configuration of the Guardian previously took 710 lines (https://github.com/wormhole-foundation/wormhole/blob/main/node/cmd/guardiand/node.go#L783-L1493). It is now reduced to 501 lines which are much more clear and less error-prone. 
    * Watchers are now configured through the `WatcherConfig` interface, which I implemented for each watcher. 

Functional changes
* *All* watchers will be run with `common.WrapWithScissors`. Before, the following ran without: Wormchain, Aptos, Sui.
* A few log messages may be changed
* Other than that, there should be no functional changes in this PR! 

Future work (out of scope)
* This PR includes a bare-bones unit test in node_test.go to demonstrate that this refactor allows writing better unit tests. In future PRs we can add more extensive unit testing and also negative testing. For example, the tests proposed in https://github.com/wormhole-foundation/wormhole/pull/2652 should be possible to better express in node_test.go. 
* The unit test in node_test.go currently runs ca. 15 seconds. This can probably be optimized. The bottleneck is p2p startup time I believe. 
# Testing and roll-out
Main concern
* When making a refactor, the best strategy is to get good test coverage for the existing code and then verify that tests are still passing after the refactor. 
* Unfortunately, there is no good test coverage of the existing code. This PR is exactly to address that. 
* The different EVM watchers are not tested in Tilt (devnet). In Tilt we only have Ethereum and BSC.

Action items
- [x] @tbjump Make a PR to change each watcher to log its config on creation/startup ("watcher_startup"): https://github.com/wormhole-foundation/wormhole/pull/3076
- [x] Get it merged to main and then rebase the node refactor over it (should have no conflicts).
- [x] Ensure that test coverage for the new code in node/node.go is close to 100% (separate PR: https://github.com/wormhole-foundation/wormhole/pull/3067)

Testing strategy
- [x] Record the `watcher_startup` logs from the latest release in testnet.
- [x] Make a pre-release with the refactor
- [x] Run the pre-release in testnet
- [x] Check that the `watcher_startup` of the pre-release are exactly the same as the ones in testnet.
- [x] Send a message from each chain in testnet and verify that a VAA was created for it
- [x] Roll-out the pre-release to one Guardian on mainnet
- [x] Verify that this Guardian is still participating in consensus on all the chains, by looking at wormscan or by looking at the logs
- [ ] Promte the release to `latest`

tbjump commented:
Current dependencies on/for this PR:
* main
  * **PR #2980** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/2980" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #2983** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/2983" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #3024** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3024" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3024?utm_source=stack-comment).

tbjump commented:
changing to draft while working on resolving the failed CI tests

panoel commented on node/cmd/guardiand/node.go:
```diff
@@ -836,33 +795,6 @@ func runNode(cmd *cobra.Command, args []string) {
-	solAddress, err := solana_types.PublicKeyFromBase58(*solanaContract)
-	sepoliaContractAddr := eth_common.HexToAddress(*sepoliaContract)
-	if err != nil {
-		logger.Fatal("invalid Solana contract address", zap.Error(err))
```
This error message gets lost in the changes.  Not sure if it is extremely important.

tbjump replied:
This check is in https://github.com/wormhole-foundation/wormhole/blob/2fdb18e29ffdc572ab5a4f222c4d8e05a9aef652/node/pkg/watchers/solana/config.go#L48 and would also generate a logger.Fatal in https://github.com/wormhole-foundation/wormhole/blob/2fdb18e29ffdc572ab5a4f222c4d8e05a9aef652/node/pkg/node/node.go#L250.

panoel commented on node/cmd/guardiand/node.go:
```diff
@@ -893,110 +818,80 @@ func runNode(cmd *cobra.Command, args []string) {
 		logger.Fatal("failed to load guardian key", zap.Error(err))
 	}
 
-	guardianAddr := ethcrypto.PubkeyToAddress(gk.PublicKey).String()
```
This seems like a simplification that would have been fine to keep so that you don't do 
```
ethcrypto.PubkeyToAddress(gk.PublicKey)
```
everywhere.

panoel commented on node/pkg/node/node.go:
```diff
@@ -0,0 +1,574 @@
+
+			chainObsvReqC := make(map[vaa.ChainID]chan *gossipv1.ObservationRequest)
+
+			go handleReobservationRequests(ctx, clock.New(), logger, g.obsvReqC.readC, chainObsvReqC)
```
This appears to be structured differently.  Previously, the watchers were started before handling the reobservation requests.  Is there a reason to change that?

tbjump replied:
The order in which these two goroutines (watcher and reobservation request handler) are started should not matter since they're running concurrently anyways.

panoel replied:
The code as-is is passing in an empty map to the go func.  I **really** think that the watchers should all be created and the chainObsvReqC map fully filled out before you call this.

panoel replied:
I think that if you moved this line down to the bottom (just before returning), it would be as before.

tbjump commented on node/cmd/guardiand/node.go:
```diff
@@ -1092,577 +984,419 @@ func runNode(cmd *cobra.Command, args []string) {
-				}
-			}
-			if shouldStart(sepoliaRPC) {
-				if ethWatcher == nil {
```
yes, this check went missing. But is it needed? Why does sepolia require `ethWatcher`?

panoel replied:
Yea, probably don't need the check.

panoel commented on node/pkg/node/adminServiceRunnable.go:
```diff
@@ -25,7 +25,7 @@ import (
 )
 
-func AdminServiceRunnable(
+func adminServiceRunnable(
```
If you are going to change this name, shouldn't it be consistent with the naming used in `node/pkg/node/guardianKey.go`?

tbjump replied:
I thought that `GuardianKeyToAddress` may be useful in other places of the codebase and it'd be a good function to export vs. `adminServiceRunnable`, which isn't and shouldn't be used elsewhere. Wdyt?

panoel replied:
Wasn't aware of the Go naming convention stuff.  That should be fine.

panoel commented on node/pkg/p2p/p2p.go:
```diff
@@ -280,14 +280,14 @@ func Run(
 
 			ctr := int64(0)
-			tick := time.NewTicker(15 * time.Second)
-			defer tick.Stop()
```
Shouldn't we keep the defer or put a tick.Stop() in the ctx.Done() case below?

tbjump replied:
good catch, this should've been changed to time.Timer. The reason to switch to `time.Timer` instead of `time.Ticker` was so that we can do the first iteration after one millisecond instead of 15 seconds, such that guardian immediately send out a heartbeat and the tests complete faster.

panoel commented:
If you wanted, the `TODO` you added in comments can be converted into issues so that they don't get lost?

tbjump commented:
Added a few more test cases and also ported over the test cases proposed in https://github.com/wormhole-foundation/wormhole/pull/2652. 

I just had a tilt run succeed locally, but CI still seems to be failing. Looking into it further.

panoel commented on node/pkg/node/node.go:
```diff
@@ -0,0 +1,575 @@
+		// run the config
+		err := option.f(ctx, logger, g)
+		if err != nil {
+			return err
```
Can we include the component with the error just in case the lower layer isn't including the component name?

tbjump replied:
done

tbjump commented:
passed ci tests in devnet (locally)

tbjump commented:
I'm working on a few more improvements (e.g. increasing test coverage of node.go). I created a new branch #3067 stacked upon this PR such that this PR remains stable for review and doesn't need to run ci tests every time I make a small change improvement.

tbjump commented:
rebased and addressed merge conflicts.

tbjump commented:
rebased and addressed merge conflicts (again)

tbjump commented:
CI Tilt tests are still passing when I run it on my machine, but not on CI. There seems to be a race condition somewhere. 
This PR changes the order in which some of the goroutines are started. This should not make any meaningful difference because they are running concurrently anyways, but it could surface previously existing race conditions and this is what may be causing Tilt tests to fail in CI. 

I have added a temporary commit to change the guardiand log level to "info" in CI. 

I am getting this log entry which I haven't seen before: 
`2023-06-14T13:52:05.3612272Z      guardian │ 2023-06-14T13:52:05.358Z	INFO	swarm2	swarm/swarm_dial.go:393	error resolving /dns4/guardian-0.guardian/udp/8999/quic/p2p/12D3KooWL3XJ9EMCyZvmmGXL2LMiVBtrVa2BuESsJiXkSj7333Jw: lookup guardian-0.guardian on 10.96.0.10:53: no such host
`

Maybe the issue is that guardian-1's p2p runner is starting up before guardian-0. I added a commit to add a 5 second delay to the startup of non-bootstrap guardians. Let's see if that fixes the issue.

tbjump commented:
That seems to have fixed it. I'll split out the race condition fix as a separate PR and then we should be good to go here.

tbjump commented on node/pkg/node/node.go:
```diff
@@ -0,0 +1,575 @@
+						case msg := <-c:
+							if msg.EmitterChain != chainId {
+
+								logger.Error("SECURITY CRITICAL: Received observation from a chain that was not marked as originating from that chain",
```
I added a test in https://github.com/wormhole-foundation/wormhole/blob/ed009d0976fe38f9a72abe8411c8746b6d735700/node/pkg/node/node_test.go#L334 that covers this code path. If it was logger.Fatal() here, then that would cause an os.Exit(), which we cannot catch in the test. We could dynamically overwrite logger.Fatal() in the test such that it doesn't call os.Exit() but I think logger.Error() is actually more appropriate here because the error is recoverable. 
The reason we did logger.Fatal() before was not because it was an irrecoverable error but because we wanted to nudge operators to take a look at and report this error. I think the correct thing to do would be to instead have operators monitor logs for "SECURITY CRITICAL".

tbjump replied:
I disagree, but I made the change as requested.

tbjump commented on node/pkg/common/readiness.go:
```diff
@@ -13,7 +13,7 @@ const (
 
 // MustRegisterReadinessSyncing registers the specified chain for readiness syncing. It panics if the chain ID is invalid so it should only be used during initialization.
-// This function will
+// TODO: Using vaa.ChainID is bad here because there can be multiple watchers for the same chainId, e.g. solana-finalized and solana-confirmed. This is currentl handled as a special case for solana in node/node.go, but should really be fixed here.
```
fixed

tbjump commented on node/pkg/node/node.go:
```diff
@@ -0,0 +1,575 @@
+				if w.RequiredL1Finalizer() != "" {
+					l1watcher, ok := watchers[w.RequiredL1Finalizer()]
+					if !ok || l1watcher == nil {
+						logger.Fatal("L1finalizer does not exist. Please check the order of the watcher configuration. The L1 must be configured before this one.",
```
The variable is called `watcherConfigs`. I changed the message to `L1finalizer does not exist. Please check the order of the watcher configurations in watcherConfigs. The L1 must be configured before this one.` does that address your concern?

tbjump commented on .github/workflows/build.yml:
```diff
@@ -318,7 +318,7 @@ jobs:
       # The go-ethereum and celo-blockchain packages both implement secp256k1 using the exact same header, but that causes duplicate symbols.
       - name: Run golang tests
-        run: cd node && go test -v -timeout 1m -race -ldflags '-extldflags "-Wl,--allow-multiple-definition" ' ./...
+        run: cd node && go test -v -timeout 5m -race -ldflags '-extldflags "-Wl,--allow-multiple-definition" ' ./...
```
The new TestConsensus tests take ~40s to run, taking the overall golang test run to slightly more than 1 minute; Most of that time is spent during libp2p startup and can probably be optimized. I'm currently working on some benchmarking/profiling to figure out where exactly the time is spent, but that shouldn't be blocking for this PR.

tbjump commented on node/pkg/common/chainlock.go:
```diff
@@ -155,3 +156,20 @@ func (msg *MessagePublication) CreateDigest() string {
+		zap.String("message_id", string(msg.MessageID())),
+		zap.Bool("unreliable", msg.Unreliable),
+	)
+}
```
The same fields are already logged most of the time, but with different names. I'm working on benchmarks that will also measure the log volume generated and we can optimize there. I think we'll primarily want to reduce the number of times log entries are generated at all vs. the details in each log entry, but we can cross that bridge when we get there.

tbjump commented on node/pkg/node/node.go:
```diff
@@ -0,0 +1,533 @@
+
+// GuardianOptionP2P configures p2p networking.
+// Dependencies: Accountant, Governor
+func GuardianOptionP2P(p2pKey libp2p_crypto.PrivKey, networkId string, bootstrapPeers string, nodeName string, disableHeartbeatVerify bool, port uint, ibcFeatures string) GuardianOption {
```
Yeah don't love the way that p2p is setup, but we can leave that for another day.

tbjump commented on node/pkg/node/node.go:
```diff
@@ -0,0 +1,575 @@
+				for _, chainID := range ibc.Chains {
+
+					if _, exists := chainMsgC[chainID]; !exists {
+						return errors.New("invalid IBC chain ID")
```
I'll add it to the follow-up PR https://github.com/wormhole-foundation/wormhole/pull/3067

tbjump commented on node/pkg/node/node.go:
```diff
@@ -0,0 +1,575 @@
+	obsvReqSendC channelPair[*gossipv1.ObservationRequest]
+	// Injected VAAs (manually generated rather than created via observation)
+	injectC channelPair[*vaa.VAA]
+	// TODO describe
```
Added a comment `// acctC is the channel where messages will be put after they reached quorum in the accountant.` Also addressed the documentation TODO a few lines above.

tbjump commented on node/pkg/node/node.go:
```diff
@@ -0,0 +1,533 @@
+					g.acctC.writeC,
+					g.env,
+				)
+			} else {
+				logger.Info("acct: accountant is disabled", zap.String("component", "gacct"))
```
changed it

tbjump commented on node/pkg/node/node.go:
```diff
@@ -0,0 +1,533 @@
+		}}
+}
+
+func GuardianOptionGovernor(governorEnabled bool) GuardianOption {
```
comments are added in https://github.com/wormhole-foundation/wormhole/pull/3067

tbjump commented on node/pkg/node/guardianKey.go:
```diff
@@ -0,0 +1,12 @@
+
+func GuardianKeyToAddress(publicKey ecdsa.PublicKey) eth_common.Address {
+	return ethcrypto.PubkeyToAddress(publicKey)
+}
```
since the function actually wasn't used that much, I just eliminated it.

SEJeff replied:
That's fine, but as a general rule. One function per file is excessive unless they're 100 LOC or more per function.

tbjump commented on node/pkg/node/node.go:
```diff
@@ -0,0 +1,533 @@
+									zap.Stringer("msgChainId", msg.EmitterChain),
+									zap.Stringer("watcherChainId", chainId),
+								)
+							} else if msg.EmitterAddress == zeroAddress {
```
I'll change it to Fatal (see also conversation w/ @bruce-riley above).

evan-gray reviewed (dismissed):
approving `.github/workflows`

tbjump commented:
[v2.18.5-test3](https://github.com/wormhole-foundation/wormhole/releases/tag/v2.18.5-test3) worked in testnet with all chains except the ones who are currently down in testnet: 
Aurora - down in testnet
Klaytn - down in testnet
Neon - testnet guardian is too far behind on this chain
Base - down in testnet
Terra Classic - RPC unavailable. 

I think this release can therefore be adopted by one mainnet guardian for verification in mainnet.

SEJeff commented on node/pkg/node/node.go:
```diff
@@ -0,0 +1,575 @@
+		}}
+}
+
+type G struct {
```
Totally understood. One letter variables, especially for important stuff, is just bad. I'll respectfully disagree with the other reviewer, but not block it :)

tbjump commented:
As of 5:10 PM UTC, the Jump Guardian is running v2.18.5-test3, which includes this refactor.

tbjump commented:
The Jump Guardian is still participating in consensus for all chains that have activity. 
I will resolve the merge conflict and then re-request approvals and then we should be good to merge this branch.

tbjump commented:
Added a PR for the script that I used: https://github.com/wormhole-foundation/wormhole/pull/3122. 

It looks at Wormscan and lists which guardians participated in the most recent VAAs. 

Output: 

```
go run guardian_vaa_stats.go 2023-06-23 05:10:00 +0000 UTC
Only looking for VAAs after 2023-06-23 05:10:00 +0000 UTC

solana	map[0:7 1:10 2:10 3:5 4:10 6:10 9:7 10:10 11:5 13:9 14:10 15:8 16:10 17:10 18:10]
ethereum	map[0:9 1:6 2:10 3:10 4:9 6:10 7:3 8:1 9:8 10:7 12:3 13:8 14:8 15:9 16:10 17:9 18:10]
terra	map[]
bsc	map[0:10 1:9 2:10 3:9 4:10 6:10 9:9 10:9 13:7 14:9 15:10 16:10 17:8 18:10]
polygon	map[0:10 1:8 2:10 3:10 5:5 6:1 7:3 9:8 10:8 11:7 12:2 13:10 14:10 15:10 16:10 17:10 18:8]
avalanche	map[0:10 1:10 2:8 3:9 4:8 5:1 6:10 7:1 8:2 9:4 10:7 11:5 12:2 13:9 14:10 15:9 16:8 17:9 18:8]
oasis	map[]
algorand	map[]
aurora	map[]
fantom	map[]
karura	map[]
acala	map[]
klaytn	map[]
celo	map[0:8 1:8 2:8 3:8 4:8 6:7 9:2 10:7 11:2 13:8 14:6 15:8 16:8 17:8 18:8]
near	map[]
moonbeam	map[0:2 1:2 2:1 3:2 4:2 6:1 9:2 10:1 11:1 13:2 14:2 15:2 16:2 17:2 18:2]
neon	map[]
terra2	map[]
injective	map[]
sui	map[0:10 1:8 2:10 3:10 4:9 6:7 7:1 9:4 10:9 11:3 13:9 14:10 15:10 16:10 17:10 18:10]
aptos	map[0:10 2:10 3:10 4:10 5:1 6:10 9:6 10:10 11:3 13:10 14:10 15:10 16:10 17:10 18:10]
arbitrum	map[0:9 1:9 2:6 3:10 4:10 5:1 6:10 7:3 8:3 9:9 10:7 11:8 12:5 13:9 15:7 16:10 17:4 18:10]
optimism	map[0:1 1:1 2:1 3:1 4:1 6:1 10:1 12:1 13:1 15:1 16:1 17:1 18:1]
pythnet	map[0:10 1:7 2:10 3:8 4:9 5:3 6:6 8:6 9:7 10:8 11:9 12:5 13:10 14:7 15:10 16:10 17:10 18:10]
xpla	map[]
btc	map[]
base	map[]
sei	map[]
wormchain	map[]
sepolia	map[]
```

As you can see, the Jump Guardian running [v2.18.5-test3](https://github.com/wormhole-foundation/wormhole/releases/tag/v2.18.5-test3) participated in consensus on all chains that have activity.

evan-gray reviewed (approved):
reviewed `.github/workflow`

tbjump commented:
Latest run of the script. Everything looks good. The Jump Guardian running [v2.18.5-test3](https://github.com/wormhole-foundation/wormhole/releases/tag/v2.18.5-test3) participated in consensus on all chains that have activity.

```
$ go run guardian_vaa_stats.go 2023-06-23 05:10:00 +0000 UTC

Only looking for VAAs after 2023-06-23 05:10:00 +0000 UTC

solana	map[0:8 1:9 2:8 3:9 4:10 6:10 8:1 9:6 10:10 11:3 12:7 13:9 14:10 15:1 16:10 17:9 18:10]
ethereum	map[0:8 1:6 2:9 3:9 4:9 5:4 6:10 7:1 9:10 10:9 11:4 12:4 13:10 14:10 15:4 16:10 17:10 18:9]
terra	map[]
bsc	map[0:9 1:9 2:10 3:9 4:9 5:2 6:10 9:5 10:9 11:4 12:2 13:10 14:10 15:2 16:10 17:10 18:10]
polygon	map[0:10 1:9 2:10 3:10 5:4 6:8 7:8 8:8 9:6 10:8 11:8 13:10 14:10 15:8 16:9 17:10 18:9]
avalanche	map[0:10 1:10 2:10 3:10 4:10 6:10 9:8 10:8 11:1 12:7 13:9 14:7 16:10 17:10 18:10]
oasis	map[]
algorand	map[0:4 1:4 2:5 3:5 4:5 5:1 6:5 7:1 8:1 9:2 10:5 11:1 12:1 13:3 14:4 15:4 16:5 17:5 18:4]
aurora	map[]
fantom	map[0:9 1:9 2:8 3:10 4:10 6:9 9:7 10:9 11:1 13:10 14:10 15:10 16:10 17:8 18:10]
karura	map[0:9 1:9 2:9 4:9 5:7 6:9 9:4 10:8 11:6 13:4 14:9 15:9 16:8 17:8 18:9]
acala	map[0:1 2:1 3:1 4:1 6:1 9:1 11:1 13:1 14:1 15:1 16:1 17:1 18:1]
klaytn	map[0:1 1:1 2:1 3:1 4:1 5:1 6:1 7:1 9:1 11:1 14:1 15:1 16:1]
celo	map[0:10 1:10 2:10 3:10 4:10 6:10 9:6 10:10 11:2 13:7 14:9 15:10 16:10 17:6 18:10]
near	map[]
moonbeam	map[0:10 1:9 2:10 3:10 4:8 5:2 6:10 9:7 10:5 11:1 12:4 13:9 14:10 15:10 16:10 17:10 18:5]
neon	map[]
terra2	map[]
injective	map[]
sui	map[0:10 1:10 2:10 3:10 4:8 6:9 9:7 10:6 13:10 14:10 15:10 16:10 17:10 18:10]
aptos	map[0:9 2:10 3:10 4:10 5:2 6:10 9:4 10:10 11:6 12:8 13:10 14:9 15:2 16:10 17:10 18:10]
arbitrum	map[0:9 1:6 2:3 3:9 4:6 5:1 6:8 7:5 8:6 9:9 10:6 11:8 12:7 13:9 14:6 15:5 16:10 17:9 18:8]
optimism	map[0:10 1:8 2:10 3:10 4:9 6:10 9:6 10:8 12:3 13:8 14:9 15:10 16:10 17:9 18:10]
pythnet	map[0:9 1:4 2:10 3:5 4:10 5:1 6:6 8:10 9:5 10:10 11:10 12:10 13:10 14:5 15:1 16:10 17:10 18:10]
xpla	map[]
btc	map[]
base	map[]
sei	map[]
wormchain	map[]
sepolia	map[]
```
